### PR TITLE
Snapshot implementation draft b2

### PR DESCRIPTION
**Pull request by nyaxt**

### CODE REVIEW

File: draft-yasskin-httpbis-origin-signed-exchanges-impl.md [hunk 1]
```diff
@@ -899,11 +916,22 @@ This content type consists of the concatenation of the following items:
    signature.
 1. 3 bytes storing a big-endian integer `sigLength`. If this is larger
```
Comment (irori) [inline draft-yasskin-httpbis-origin-signed-exchanges-impl.md]: Remove this and next line

Comment (nyaxt) [reply draft-yasskin-httpbis-origin-signed-exchanges-impl.md]: Done.

File: draft-yasskin-httpbis-origin-signed-exchanges-impl.md [hunk 2]
```diff
@@ -175,9 +175,12 @@ values:
 
 "integrity"
 
-: A string (Section 3.7 of {{!I-D.ietf-httpbis-header-structure}}) containing
-  the lowercase name of the response header field that guards the response
-  payload's integrity.
+: A string (Section 3.7 of {{!I-D.ietf-httpbis-header-structure}}) containing a
+  "/"-separated sequence of names starting with the lowercase name of the
```
Comment (twifkak) [inline draft-yasskin-httpbis-origin-signed-exchanges-impl.md]: May not be worth deciding now, but future integrity formats may want names containing a variable number of slashes (e.g. a pair of base64-encoded strings), so you may want an escaping mechanism eventually.

File: draft-yasskin-httpbis-origin-signed-exchanges-impl.md [hunk 3]
```diff
@@ -461,13 +460,16 @@ to retrieve an updated OCSP from the original server.
    certificate chain in its Certificate Verify (Section 4.4.3 of
    {{?I-D.ietf-tls-tls13}}), in order to allow updating the stapled OCSP
    response without updating signatures at the same time.
-1. If `signature` is a valid signature of `message` by `publicKey` using
-   `signing-alg`, return "potentially-valid" with `certificate-chain`.
-   Otherwise, return "invalid".
+1. If `signature` is not a valid signature of `message` by `publicKey` using
+   `signing-alg`, return "invalid".
+1. If `integrity` does not match "digest/mi-sha256-03", return "invalid".
```
Comment (twifkak) [inline draft-yasskin-httpbis-origin-signed-exchanges-impl.md]: Why move this down here from line 419?

File: draft-yasskin-httpbis-origin-signed-exchanges-impl.md [hunk 4]
```diff
@@ -890,20 +907,29 @@ and header fields, and a response payload.
...
    Note: As this is a snapshot of a draft of
    {{?I-D.yasskin-http-origin-signed-responses}}, it uses a distinct file
    signature.
-1. 3 bytes storing a big-endian integer `sigLength`. If this is larger
-   than 16kB, parsing MUST fail.
+1. 2 bytes storing a big-endian integer `fallbackUrlLength`.
+1. `fallbackUrlLength` bytes holding a `fallbackUrl`, which MUST be an absolute
+   URL with a scheme of "https".
+
+   Note: The byte location of the fallback URL is intended to remain invariant
```
Comment (twifkak) [inline draft-yasskin-httpbis-origin-signed-exchanges-impl.md]: This means the magic "sxg1-b2" must always remain 7 chars. Out of curiosity:
 - What's after `b9`? `bA`..`bZ`? Or do we move to `g1` (for gamma)? :P
 - What's v1 look like? "sxg1\0\0\0"?

File: draft-yasskin-httpbis-origin-signed-exchanges-impl.md [hunk 5]
```diff
@@ -890,20 +907,29 @@ and header fields, and a response payload.
...
+1. The ASCII characters "sxg1-b2" followed by a 0 byte, to serve as a file
    signature. This is redundant with the MIME type, and recipients that receive
    both MUST check that they match and stop parsing if they don't.
 
    Note: As this is a snapshot of a draft of
    {{?I-D.yasskin-http-origin-signed-responses}}, it uses a distinct file
    signature.
-1. 3 bytes storing a big-endian integer `sigLength`. If this is larger
-   than 16kB, parsing MUST fail.
+1. 2 bytes storing a big-endian integer `fallbackUrlLength`.
```
Comment (twifkak) [inline draft-yasskin-httpbis-origin-signed-exchanges-impl.md]: (Given 65536 is longer than several existing browsers can support (AFAICT), perhaps the fetch spec should say how browsers should respond if the fallbackUrl is too long? Or specify a minimum length they must support? Or is this already covered by some other part of Fetch?)

File: draft-yasskin-httpbis-origin-signed-exchanges-impl.md (hunk 2 above)
Comment (nyaxt) [reply draft-yasskin-httpbis-origin-signed-exchanges-impl.md]: IIUC, the string here refers to format id such as "digest/mi-sha256-03" so I think we are OK?

File: draft-yasskin-httpbis-origin-signed-exchanges-impl.md (hunk 3 above)
Comment (nyaxt) [reply draft-yasskin-httpbis-origin-signed-exchanges-impl.md]: We wanted to clarify that the UAs may start processing the headers part before verifying the entire payload.
Before this change, invalid payload could make headers invalid. After this change the headers checks are somewhat independent.

The original change was from @jyasskin, so I may be wrong.

File: draft-yasskin-httpbis-origin-signed-exchanges-impl.md (hunk 4 above)
Comment (nyaxt) [reply draft-yasskin-httpbis-origin-signed-exchanges-impl.md]: > - What's after b9? bA..bZ? Or do we move to g1 (for gamma)? :P

We hope to get to a standard before reaching bA :airplane: 

> - What's v1 look like? "sxg1\0\0\0"?

That is the idea.

File: draft-yasskin-httpbis-origin-signed-exchanges-impl.md (hunk 5 above)
Comment (nyaxt) [reply draft-yasskin-httpbis-origin-signed-exchanges-impl.md]: I'm not sure. Would you mind filing a separate issue?

File: draft-yasskin-httpbis-origin-signed-exchanges-impl.md (hunk 2 above)
Comment (twifkak) [reply draft-yasskin-httpbis-origin-signed-exchanges-impl.md]: Yeah, I think this is fine for the foreseeable future; I was just hypothesizing about a new integrity header.

File: draft-yasskin-httpbis-origin-signed-exchanges-impl.md (hunk 5 above)
Comment (twifkak) [reply draft-yasskin-httpbis-origin-signed-exchanges-impl.md]: OK, filed #297.

**Comment by nyaxt** [comment]
Updated ChangeLog. PTAL

File: draft-yasskin-httpbis-origin-signed-exchanges-impl.md [hunk 6]
```diff
@@ -1018,6 +1044,23 @@ The other fields are the same as the registration in
...
 
+draft-02
+
+Vs. draft-01:
+
+* Define absolute URLs, and limit the schemes each instance can use.
+* Update to mice-03 including the Digest header.
+* Define the "integrity" field of the Signature header to include the digest
+  algorithm.
+* Put a fallback URL at the beginning of signed exchanges, and remove ':url' key
```
[the remark this answers is not in the document]
Comment (nyaxt) [reply draft-yasskin-httpbis-origin-signed-exchanges-impl.md]: Done.

**Pull request merged**